docs/source/contributing.rst

```diff
@@ -1,12 +1,210 @@
+
+If the reason for the failure isn't obvious or it's a problem in the CI check itself, seek some help.
+
+**Content**
+
+It's important that the content of a pull request be kept clean and small. Pull
+requests should be less than 1-2k lines of code. The code changes should
+reflect one and only one topic (e.g do not include two bug fixes in one pull
+request). Content should generally be kept to code and documentation, binary
+content, such as images, may be uploaded elsewhere.
+
+The description of the PR should accurately describe what the purpose of the changes are. The associated issue should link any PR or TODOs that are being addressed. A large refactor will require design documents to be written. Steps to test is a bulleted list that helps your reviewers on how to test your PR.
+
+Docstrings
+~~~~~~~~~~
+Code should be documented thoroughly. Generally speaking, you won't be here for
+more than 4(ish) years. Many students will come behind you and will need to use
+the code you've written.
+
+Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``) to HTML files. This build process is done by Sphinx. The Python-side API is done by a Sphinx extension named ``sphinx-autodoc``.
+
+.. note::
+
+    As of the time of this writing, there is no C++-side API docs yet, but in theory Sphinx + Doxygen + Breathe allow this functionality.
+
+When writing documentation for your Python code, it is very important to follow a guide so the ``sphinx-autodoc`` can accurately create documentation which is found in the "API Reference" page.
+
+In general, a typical ``Sphinx`` docstring has the following format:
+
+.. code-block:: python
+
+    """[Summary]
+
+    :param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
+    :type [ParamName]: [ParamType](, optional)
+    ...
+    :raises [ErrorType]: [ErrorDescription]
+    ...
+    :return: [ReturnDescription]
+    :rtype: [ReturnType]
+    """
+
+Each function should have documentation containing it's purpose, what
+parameters is takes, and what values it returns. Error handling should be
+described as well: what errors will it produce, and what assumptions it makes
+regarding the validation the caller performs ahead of time.
+
+If the overall set of code is complex and new, consider editing or adding to
+the docs.
+
+For more information on writing Sphinx docstrings, you can learn more `here <https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html>`_.
+
+Text file types that are not supported by doxygen or Sphinx should still be documented using what ever commenting style that format supports.
+
+**Style and Formatting**
+
+In order to keep the code more readable, code should be formatted and styled
+uniformly. This would be difficult to coordinate across multiple users, so we
+have a program that automatically restyles the code for you. Our precommit hook runs ``make pretty-lines`` which auto-formats the code.
```

Comment:
actually right now pretty-lines precommit hook is a little broken, can you add that as a card